Tutorial from Manon but cute (#33)
scilus/nf-neuro-documentation: src/content/docs/pipeline/tutorial_step_6.mdx

        
        script:
        def prefix = task.ext.prefix ?: "${meta.id}"
        def optionName = task.ext.option ? "${task.ext.option}" : "default_value"
        """
        cat ${input_file} > output.txt
        """
    }
    ```

    :::tip[Let's break down these changes]
    <html>
    <li>`tag "$meta.id"`</li>
        This assigns a tag to the process using the 'id' field from the 'meta' map. 
        Tags are useful for logging and tracking purposes.

    <li>`label 'process_single'`</li>
        This assigns a label to the process, which can be used for 
        resource allocation or other configuration purposes.

    <li>`container`</li>
        Specifies the container to use for the process. 
        It conditionally chooses between a Singularity or Docker container based on the workflow configuration.

    <li>`Input`</li>
        Defines the input for the process as a tuple containing 
        metadata (meta) and an input file path.

    <li>`Output`</li>
        Defines the output of the process as a tuple containing 
        the input metadata and a file named "output.txt". 
        The output is emitted to a channel named "outputname".

    <li>`script`</li>

>         This is the script that will be executed by the process. 
>         It defines some option a `prefix` variable using a `task` extension or the meta id.
>         Defines an 'optionName' variable using a task extension or a default value.

Reviewer: ```suggestion
        This is the section in which the command to be executed by the module are defined. 
        It is also where you can define options, such as a `prefix` variable using a `task` extension or the `meta.id`.
        You can also defines optional parameters, such as the 'optionName' variable using a task extension or a default value.
```

    </html>
    :::
    
    Here it, execute a simple command to copy the contents of the input file to output.txt.
    In summary, this process takes an input file, copies its contents to an output file,
    and provides some flexibility in naming and options through the use of metadata and task extensions. 
    The process can run in either a Singularity or Docker container, depending on the workflow configuration.

    ### 3. Write the local module

    The goal of this local module is to segment a T1-weighted (T1w) image using `fast` and 
    extract metrics from the white matter (WM), gray matter (GM), and cerebrospinal fluid (CSF) 
    masks using the `scil_volume_stats_in_ROI.py` script from the scilpy toolbox.

    You can copy and paste the example module into your `main.nf` file. First, rename the process 
    `STATS_METRICSINROI` and then modify it using the steps below:

    <Steps>
    
    1. Modify `input` to include T1 and metrics
        ```nextflow
        tuple val(meta), path(t1), path(metrics)
        ```

>     2. Update `output` to include segmentation maps and masks, as well as the JSON output from scilpy script
>         Example for WM data:

Reviewer: ```suggestion
    2. Update `output` to include segmentation maps and masks, as well as the JSON output from the scilpy script.
```

        ```nextflow
            output:
            tuple val(meta), path("*.json")                         , emit: stats
            tuple val(meta), path("*mask_wm.nii.gz")                , emit: wm_mask
            tuple val(meta), path("*map_wm.nii.gz")                 , emit: wm_map

        ```

>     3. Include `task` extensions for additional options (those from scilpy script)

Reviewer: ```suggestion
    3. Include `task` extensions for additional options (those from the scilpy script)
```

        ```nextflow
            script:
            def prefix = task.ext.prefix ?: "${meta.id}"
            def bin = task.ext.bin ? "--bin " : ""
            def normalize_weights = task.ext.normalize_weights ? "--normalize_weights " : ""
        ```

    4. Modify the `script`

>             1. Update the script to perform segmentation using `fast` on the T1w image.
> 
>             2. Generate binary masks for each PVE tissue segmentation using `scil_volume_math.py` or another tool.
>             
>             3. Rename PVE outputs from `fast` to map_\*.nii.gz and mask_*.nii.gz
> 
>             - Extract metrics from the WM, GM and CSF binary masks using `scil_volume_stats_in_ROI.py` or other.

Reviewer: ```suggestion
        <Steps>
            1. Update the script section to perform segmentation using `fast` on the T1w image.

            2. Generate binary masks for each PVE tissue segmentation using `scil_volume_math.py` or another tool.
            
            3. Rename PVE outputs from `fast` to `map_\*.nii.gz` and `mask_*.nii.gz`.

            4. Extract metrics from the WM, GM and CSF binary masks using `scil_volume_stats_in_ROI.py` or other.
        </Steps>
```

    export OPENBLAS_NUM_THREADS=1

    fast -t 1 -n 3\
        -H 0.1 -I 4 -l 20.0 -g -o t1.nii.gz $t1
    scil_volume_math.py convert t1_seg_2.nii.gz ${prefix}__mask_wm.nii.gz --data_type uint8
    scil_volume_math.py convert t1_seg_1.nii.gz ${prefix}__mask_gm.nii.gz --data_type uint8
    scil_volume_math.py convert t1_seg_0.nii.gz ${prefix}__mask_csf.nii.gz --data_type uint8
    mv t1_pve_2.nii.gz ${prefix}__map_wm.nii.gz
    mv t1_pve_1.nii.gz ${prefix}__map_gm.nii.gz
    mv t1_pve_0.nii.gz ${prefix}__map_csf.nii.gz

    scil_volume_stats_in_ROI.py ${prefix}__mask*.nii.gz  \
        --metrics $metrics \
        --sort_keys \
        $bin $normalize_weights > ${prefix}__stats.json

    """
    }
```

### 4. Use and prepare the input structure of the local module in the pipeline

#### a. Use your local module

You can now include and use your local module to the workflow, as shown in the previous steps:

```nextflow
// Add this line on the top of the main.nf
include { STATS_METRICSINROI } from './modules/local/stats/metricsinrois/main'

// Add the local module in the workflow
STATS_METRICSINROI( input_channel )
```

#### b. Prepare the input channel for your local module

> The STATS_METRICSINROIS module requires two inputs: a T1 image and a list of metrics.

Reviewer: ```suggestion
The `STATS_METRICSINROIS` module requires two inputs: a T1 image and a list of metrics.
```

You can use the `join()` and `map()` operators to create an input channel that combines the T1 BET image output from the 
`PREPROC_T1` subworkflow with the FA map from the DTI module.

Add the input channel to the module.

**Expected input channel:**

```nextflow
input_extract_metric = PREPROC_T1.out.image_bet
            .join(RECONST_DTIMETRICS.out.fa)
            .map{ it }

STATS_METRICSINROI( input_extract_metric )
```

#### 5. Configure your local module

> To configure your local module, you need to add each task.ext.* to the module. 
> As we’ve seen previously, use the `withName` process selector to link the `ext` 
>     parameter defined in your local module to `nextflow.config` file.
> 

Reviewer: ```suggestion
To configure your local module, you need to add each `task.ext.*` to the module. 
As we’ve seen previously, use the `withName` process selector to link the `ext` 
    parameter defined in your local module to `nextflow.config` file.

```

                            .map{ it + [[]] }
            DENOISING_MPPCA( input_dwi_denoise )

            // Fetch specific output
            ch_dwi_denoised = DENOISING_MPPCA.out.image

            // Input DTI update with DWI denoised output
            input_dti_denoised = ch_dwi_denoised
                    .join(ch_dwi_bvalbvec.bvs_files)
                    .map{ it + [[]] }

            // DTI-derived metrics
            RECONST_DTIMETRICS( input_dti_denoised )

            // Preprocessing T1 images
            //inputs.anat.view()

            PREPROC_T1(
                inputs.anat,
                Channel.empty(),
                Channel.empty(),
                Channel.empty(),
                Channel.empty(),
                Channel.empty(),
                Channel.empty()
            )

            // Extract FA value 
            input_extract_metric = PREPROC_T1.out.image_bet
                    .join(RECONST_DTIMETRICS.out.fa)
                    .map{ it }
            //input_extract_metric.view()
            
            STATS_METRICSINROI( input_extract_metric )
        }
    ```
    </TabItem>
    <TabItem label="Expected meta.yml">

>     </TabItem>    

Reviewer: Expected meta is empty.
